File: src/main/java/dev/sky_lock/pocketlifevehicle/click/CarClick.java
package dev.sky_lock.pocketlifevehicle.click;

import dev.sky_lock.pocketlifevehicle.Permission;
import dev.sky_lock.pocketlifevehicle.packet.ActionBar;
import dev.sky_lock.pocketlifevehicle.util.Profiles;
import dev.sky_lock.pocketlifevehicle.vehicle.Car;
import dev.sky_lock.pocketlifevehicle.vehicle.CarEntities;
import dev.sky_lock.pocketlifevehicle.vehicle.SeatArmorStand;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;

import java.util.UUID;

/**
 * @author sky_lock
 */

public class CarClick {
    private final Player player;
    private final SeatArmorStand.CraftSeat seat;
    private final Car car;

    public CarClick(Player player, SeatArmorStand.CraftSeat seat) {
        this.player = player;
        this.seat = seat;
        this.car = CarEntities.getCar(seat.getHandle());
    }

    public void accept() {
        if (!seat.getPassengers().isEmpty()) {
            return;
        }
        CarEntities.getOwner(car).ifPresent(owner -> {
            UUID clicked = player.getUniqueId();
            String ownerName = Profiles.getName(owner);
            if (player.isSneaking()) {
                if (!clicked.equals(owner) && !Permission.CAR_CLICK.obtained(player)) {
                    sendFailureInfo("この乗り物は " + ownerName + " が所有しています");
                    return;
                }
                car.openMenu(player);
                return;
            }
            if (clicked.equals(owner)) {
                this.seat.setPassenger(player);
                return;
            }
            if (car.getStatus().isLocked()) {
                sendFailureInfo("この乗り物には鍵が掛かっています");
                return;
            }
            this.seat.setPassenger(player);
        });
    }

    private void sendFailureInfo(String message) {
        ActionBar.sendPacket(player, ChatColor.YELLOW + "" + ChatColor.BOLD + "⚠⚠ " + ChatColor.RED + "" + ChatColor.BOLD + message + ChatColor.YELLOW + "" + ChatColor.BOLD + " ⚠⚠");
    }
}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/click/CarClick.kt
package dev.sky_lock.pocketlifevehicle.click

import dev.sky_lock.pocketlifevehicle.Permission
import dev.sky_lock.pocketlifevehicle.extension.chat.plus
import dev.sky_lock.pocketlifevehicle.util.Profiles.getName
import dev.sky_lock.pocketlifevehicle.vehicle.Vehicle
import dev.sky_lock.pocketlifevehicle.vehicle.ModelArmorStand
import dev.sky_lock.pocketlifevehicle.vehicle.VehicleEntities.getCar
import dev.sky_lock.pocketlifevehicle.vehicle.VehicleEntities.getOwner
import dev.sky_lock.pocketlifevehicle.vehicle.SeatArmorStand
import org.bukkit.ChatColor
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftArmorStand
import org.bukkit.entity.Player

/**
 * @author sky_lock
 */
class CarClick(private val player: Player, private val armorStand: CraftArmorStand) {
    fun accept() {
        val handle = armorStand.handle
        val clicked = player.uniqueId
        val vehicle: Vehicle
        if (handle is SeatArmorStand) {
            if (armorStand.passengers.isNotEmpty()) {
                return
            }
            vehicle = getCar(handle) ?: return
            val owner = getOwner(vehicle) ?: return
            val ownerName = getName(owner)

            if (player.isSneaking) {
                if (clicked != owner && !Permission.VEHICLE_OPEN_GUI.obtained(player)) {
                    sendFailureInfo("この乗り物は $ownerName が所有しています")
                    return
                }
                vehicle.openMenu(player)
                return
            }

            if (clicked == owner) {
                armorStand.addPassenger(player)
                return
            }
            if (vehicle.status.isLocked) {
                sendFailureInfo("この乗り物には鍵が掛かっています")
                return
            }
            armorStand.addPassenger(player)
        } else if (handle is ModelArmorStand) {
            vehicle = getCar(handle) ?: return
            val owner = getOwner(vehicle) ?: return
            val ownerName = getName(owner)

            if (player.isSneaking) {
                if (clicked != owner && !Permission.VEHICLE_OPEN_GUI.obtained(player)) {
                    sendFailureInfo("この乗り物は $ownerName が所有しています")
                    return
                }
                vehicle.openMenu(player)
                return
            }
            if (vehicle.passengers.size >= vehicle.model.spec.capacity.value()) {
                sendFailureInfo("この乗り物は満員です")
                return
            }
            if (clicked == owner) {
                vehicle.addPassenger(player)
                return
            }
            if (vehicle.status.isLocked) {
                sendFailureInfo("この乗り物には鍵が掛かっています")
                return
            }
            vehicle.addPassenger(player)
        }

    }

    private fun sendFailureInfo(message: String) {
        player.sendActionBar(ChatColor.YELLOW + ChatColor.BOLD + "⚠⚠ " + ChatColor.RED + "" + ChatColor.BOLD + message + ChatColor.YELLOW + "" + ChatColor.BOLD + " ⚠⚠")
    }

}